KETTLE-66: Provide a standard Kettle middleware wrapper for Multer

waharnum opened the pull request:
I'm at the point where the code works and I need feedback on the API and overall design in order to finalize those so I can write some documentation both for the new middleware and the use of https://www.npmjs.com/package/form-data as a means of testing it.

Running the tests:
```
node tests/MulterTests.js
```

amb26 commented on lib/KettleMiddleware-Multer.js:
```diff
@@ -0,0 +1,149 @@
+        },
+        // Override this function for different behaviour in the file filter
+        // function
+        "getFileFilter": {
```
These two are a bit peculiar and indirect - the invoker can simply be fileFilter: {
    func: "kettle.npm.multer.allowAll"
}
etc.

waharnum replied:
Digging in further, this actually didn't even work at all - Multer doesn't export its own allowAll function (https://github.com/expressjs/multer/blob/80ee2f52432cc0c81c93b03c6b0b448af1f626e5/index.js#L7), but the issue was invisible because a "null" function for the file filter wasn't registered by Multer as a problem, as it defaulted to its internal allowAll (the one it doesn't export): https://github.com/expressjs/multer/blob/80ee2f52432cc0c81c93b03c6b0b448af1f626e5/index.js#L22

Middleware integration is fun! I'll clean this up and try to be less indirect and peculiar.

waharnum replied:
I think I've accomplished this refactoring, which has reduced the size of the file and the number of functions considerably. Thanks for the suggestions! My only uncertainty is described at https://github.com/waharnum/kettle/blob/KETTLE-66/lib/KettleMiddleware-Multer.js#L46-L58 - is this the best way to accomplish what's needed? (passing in a keyed configuration object to a third-party library's function)

amb26 commented on lib/KettleMiddleware-Multer.js:
```diff
@@ -0,0 +1,149 @@
+        },
+        // Override this function for different behaviour in the destination
+        // when using disk storage
+        "getDiskStorageDestinationFunc": {
```
As above, unwrap these by one level and make the invoker into the thing itself.

amb26 commented on lib/KettleMiddleware-Multer.js:
```diff
@@ -0,0 +1,149 @@
+};
+
+kettle.middleware.multer.getDiskStorage = function (destinationFunc, filenameFunc) {
+    var diskStorage = kettle.npm.multer.diskStorage({
```
Since this one is a "thing" it should become a member definition (in an expander) rather than an invoker when unwrapped

amb26 commented on lib/KettleMiddleware-Multer.js:
```diff
@@ -0,0 +1,149 @@
+
+kettle.middleware.multer.createMiddleware = function (that, middlewareOptions, getStorage, getMiddlewareForFileStrategy, getFileFilter) {
+
+    middlewareOptions.storage = getStorage();
```
See as above - shift this construction into a member

amb26 commented on tests/shared/DataSourceTestUtils.js:
```diff
@@ -21,15 +22,20 @@ kettle.loadTestingSupport();
+// empties a directory, used by the dataSource and multer tests when testing
+// file system writes / uploads
+// directoryPath: a path statement to a directory, resolved by fluid.module.resolvePath
+// TODO: this has replaced the concrete
```
Yes, this is perfectly fine, "TODO" comment can be axed

amb26 commented on lib/test/KettleTestUtils.http.js:
```diff
@@ -51,9 +54,86 @@ fluid.defaults("kettle.test.request", {
+// Can be used to send both files and fields
+// Used currently for testing the multer-based middleware
+//
+// TODO: Should elements of this and kettle.test.request.http be merged?
```
Yes, please derive this from kettle.test.request.http, and make the following refactoring of kettle.test.request.http's antediluvian workflow:
i) convert the "send" invoker into an event named "send", which accepts an options object rather than the broken-out five arguments
i) a) Leave the existing invoker as a backwards-compatible driver that just forwards to a firing of the event
ii) take the body of the existing "send" invoker and split it into two namespaced listeners
a) prepareRequest containing everything up to and including binding req.on("error") excluding the payload, and give this a priority of "before:sendPayload"
b) sendPayload including the preparation of the payload and sending it to the request

Then your formData grade just consists of an overriding of the sendPayload listener, together with a use of the "alternative submission method" listed here https://github.com/form-data/form-data#alternative-submission-methods in order to reuse the already created "nativeRequest"

amb26 commented on lib/test/KettleTestUtils.http.js:
```diff
@@ -51,9 +54,86 @@ fluid.defaults("kettle.test.request", {
+    // Append all files
+    fluid.each(requestOptions.formData.files, function (filePath, fieldKey) {
+        // Single file path
+        if (typeof filePath === "string") {
```
Easier is
var filePaths = fluid.makeArray(filePath);
filePaths.forEach(function (filePath) {
    var resolved = fluid.module.resolvePath(filePath);
    form.append(fieldKey, fs.createReadStream(resolved));
});

amb26 commented:
Thanks so much for the work on this, @waharnum - it's a really great addition to Kettle's functionality. Sorry to take so long to getting round to the review, haven't had time to turn back to Kettle for a while.

waharnum commented:
Thanks @amb26 for all the astute comments - I'll review and refactor soon.

waharnum commented on tests/MulterTests.js:
```diff
@@ -0,0 +1,392 @@
+                method: "POST",
+                formData: {
+                    files: {
+                        "file": kettle.tests.multer.testFilePaths.testTxtPath
```
I ended up putting them onto a holder component on the testDef structure, since they need to be reused multiple times, and referring to them via IoC references. Acceptable?

waharnum commented:
@amb26 I think I've addressed all your comments above; I'm going to start writing documentation for the new middleware and test utility grades (I think the public API should now be mostly stable, even if internal implementations change). Thanks for the review, code is much improved now.

amb26 commented on lib/KettleMiddleware-Multer.js:
```diff
@@ -0,0 +1,153 @@
+                "args": ["{that}.diskStorageConfigurationObject"]
+            }
+        },
+        "memoryStorage": {
```
It's odd to always create all of these in case any are required. A better approach would be to turn "storage" into a subcomponent, and for the polymorphism to be based on type - the user would supply
````
components: {
    storage: {
        type: "kettle.multer.diskStorage"
    }
````
etc.
Everything relating to "diskStorage" including the destination and filename go onto this subcomponent.
It would help the user if these strange callback-style pieces of configuration were unwrapped since they are really just pieces of data.
That is, the subcomponent should simply support a plain option which holds the destination, and then as a more private piece of implementation the destination function reads
kettle.middleware.multer.defaultDiskStorageDestination = function (req, file, destination, cb) {
    cb(null, destination);
};
written as an invoker so:

````
       "diskStorageDestination": {
            "funcName": "kettle.middleware.multer.defaultDiskStorageDestination",
            "args": ["{arguments}.0", "{arguments}.1", "{that}.options.destination", "{arguments}.2"]
        }
````
In this way the user can specify the destination in the expected way by simply overriding the value of the "destination" option

waharnum replied:
Agree this is better for common simple uses (including our own to date in the downstream project dependent on this work), will implement.

There are more complex use-cases that Multer supports for computing a destination, but a user can override the invoker in that case.

amb26 commented on package-lock.json:
```diff
@@ -0,0 +1,6148 @@
+{
```
Our policy is to not commit these as per the current .gitignore in trunk - https://github.com/fluid-project/kettle/blob/master/.gitignore#L3 - perhaps this branch needs merging up?

waharnum replied:
It was merged up, but there a commit of `package-lock.json` from before the current .gitignore - I've removed it.

amb26 commented on lib/KettleMiddleware-Multer.js:
```diff
@@ -0,0 +1,153 @@
+fluid.defaults("kettle.middleware.multer", {
+    gradeNames: "kettle.plainAsyncMiddleware",
+    // See https://github.com/expressjs/multer#multeropts;
+    // because Multer expects to receive functions for some
```
Although it expects to receive functions we should do our best to unwrap these back to plain data as a courtesy for implementors

waharnum commented:
@amb26 Thanks for reviewing again - I've done another round of refactoring that I hope should address thecomments about courtesy to implementors.

What I've done at a high level:
- extracted both the `storage` and `fileFilter` functionality (the areas where Multer expects functions with specific signatures as arguments when constructing the middleware) to subcomponent grades that the main Multer middleware grade now uses
- removed any expansion-based `members` except the seemingly necessary `diskStorageConfigurationObject` at line 199
- given `kettle.middleware.multer.storage.disk` plain options for `destination` and `filename` that use `fluid.stringTemplate` to interpolate values from Multer's [file information](https://github.com/expressjs/multer#file-information) object in the internal implementation - this should allow some power and flexibility in computing destination and filename without having to write new functions
- documented via comments how an advanced implementation would need to override the existing invokers

I think this is a lot cleaner now.

amb26 commented on lib/KettleMiddleware-Multer.js:
```diff
@@ -0,0 +1,229 @@
+        // object to an invoker where the object uses IoC references
+        // for the values of its keys?
+        diskStorageConfigurationObject: {
+            destination: "{that}.destination",
```
Some kind of big problem here since both of these resolve to `undefined` since they should be referenced as {that}.options.destination. Suggests that the actual values here are never tested, and only the values supplied to the invokers are important. In that case it should probably be removed. If it is retained, it's ok for it to be encoded within "options" as with its source values, but somehow it should receive dedicated tests if it is kept.

waharnum replied:
This is confusing because of bad naming (which I'll correct) - `{that}.options.destination` (a configuration value) and `{that}.destination` (an invoker) are both valid and separate, but have identical names. I'll change the invokers to `destinationImpl` and `filenameImpl` or similar so this is clearer.

Multer's own `diskStorage` function expects two functions in a key-value structure like so:

```
{
  destination: function(req, file, cb)...,
  filename: function(req, file, cb)...
}
```

So the option values are used as arguments to the invoker functions (which use argument-shifting to provide the three-value signatures Multer requires), and the invoker functions are referenced in the `diskStorageConfigurationObject`, which is passed into the actual Multer `diskStorage` function by the `multerStorage` invoker.

I couldn't find an invoker syntax that would allow me to pass a configuration object with IoC-resolved values directly, hence the storage

I find it confusing explaining this and I wrote the wrapping code, but it's how Multer's internals work - the implementation should hide this from an end user and allow most use cases to simply set `filename` and `destination` to string values with `fluid.stringTemplate` syntax.

amb26 replied:
Hi Alan, sorry to misread the code, but I guess this does highlight that it could be confusing. Perhaps "destinationResolver" or "destinationCallback" would be clearer in terms of conveying the different roles of the various incarnations of these values.

I'm not quite sure what you mean by "an invoker syntax that would allow me to pass an object with IoC-resolved values directly" - but perhaps you just mean something like this?

````
        multerStorage: {
            funcName: "kettle.npm.multer.diskStorage",
            "args": [{
                destination: "{that}.destinationResolver",
                filename: "{that}.filenameResolver"
             }]
        }
````

This will let you do without the indirection onto `diskStorageConfigurationObject`. If this wasn't what you were expecting, perhaps you could list a couple of the things you tried that didn't seem to work?

amb26 commented on lib/KettleMiddleware-Multer.js:
```diff
@@ -0,0 +1,229 @@
+    }
+});
+
+// Storage grade using Multer's diskStorage
```
These comments have received painfully short line lengths - I think the community considered that 110 or so was a reasonable line length

waharnum replied:
I normally break at 80 characters. Not sure what was going on to make me decide that 40-50 was more reasonable here. Fixed.
